gpt_index/indices/base.py

```diff
@@ -90,31 +80,64 @@ def __init__(
         self._prompt_helper = prompt_helper or PromptHelper.from_llm_predictor(
             self._llm_predictor, chunk_size_limit=chunk_size_limit
         )
-        self._text_splitter = text_splitter or self._build_fallback_text_splitter()
 
-        # build index struct in the init function
         self._docstore = docstore or DocumentStore()
         self._index_registry = index_registry or IndexRegistry()
-
         self._llama_logger = llama_logger or LlamaLogger()
+        self._node_parser = node_parser or SimpleNodeParser()
 
-        if index_struct is not None:
+        if nodes is not None: 
+            docstore.add_documents(nodes)
+            index_struct = self.build_index_from_nodes(nodes)
             if not isinstance(index_struct, self.index_struct_cls):
                 raise ValueError(
-                    f"index_struct must be of type {self.index_struct_cls}"
+                    f"index_struct must be of type {self.index_struct_cls} but got {type(index_struct)}"
                 )
-            self._index_struct = index_struct
-        else:
-            documents = cast(Sequence[DOCUMENTS_INPUT], documents)
-            documents = self._process_documents(
-                documents, self._docstore, self._index_registry
-            )
-            self._validate_documents(documents)
-            # TODO: introduce document store outside __init__ function
-            self._index_struct = self.build_index_from_documents(documents)
+
+        self._index_struct = index_struct
         # update index registry and docstore with index_struct
         self._update_index_registry_and_docstore()
 
+    @classmethod
+    def from_documents(
+        cls,
+        documents: Sequence[Document],
+        docstore: Optional[DocumentStore] = None,
+        node_parser: Optional[NodeParser] = None,
+        **kwargs: Any,
+    ) -> "BaseGPTIndex":
+        node_parser = node_parser or SimpleNodeParser()
+        docstore = docstore or DocumentStore()
+
+        # documents = cast(Sequence[DOCUMENTS_INPUT], documents)
+        # documents = self._process_documents(
+        #     documents, self._docstore, self._index_registry
+        # )
+        # self._validate_documents(documents)
+
+        # TODO: why do we do this?
```

Comment:
to allow for `refresh` capability so if user presents a set of new docs, we update only the docs that have changed in content